{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNA-Seq Workflow by @furkanmtorun \n",
    "### [furkanmtorun@gmail.com](mailto:furkanmtorun@gmail.com)  | GitHub: [@furkanmtorun](https://github.com/furkanmtorun)  | [Google Scholar](https://scholar.google.com/citations?user=d5ZyOZ4AAAAJ) | [Personal Website](https://furkanmtorun.github.io/) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Libraries , packages and required functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# +--------------------------------------------------+\n",
    "# Import required librarys & packages\n",
    "# +--------------------------------------------------+\n",
    "import glob2\n",
    "import subprocess\n",
    "\n",
    "# +--------------------------------------------------+\n",
    "# Define folders and bin for tools\n",
    "# +--------------------------------------------------+\n",
    "fastq_folder, genome_folder, index_folder, bam_sam_folder, logs_folder, results_folder, \\\n",
    "FastQC_bin, STAR_bin, cufflinks_bin, bowtie_bin, TopHat_bin, R_bin = [\"./files/fastq/\", \n",
    "            \"./files/genome/\", \"./files/index/\", \"./files/bam_sam/\", \"./files/logs/\", \"./files/results/\", \"./softs/FastQC/\", \n",
    "            \"./softs/STAR-2.7.3a/bin/Linux_x86_64/\", \"./softs/cufflinks-2.2.1.Linux_x86_64/\", \n",
    "            \"./softs/bowtie2-2.3.5.1-linux-x86_64/\", \"./softs/tophat-2.1.1.Linux_x86_64/\", \n",
    "            \"C:/Program Files/R/R-3.6.1/bin/i386/\"]\n",
    "\n",
    "\n",
    "# +--------------------------------------------------+\n",
    "# Define files\n",
    "# +--------------------------------------------------+\n",
    "fasta_files = \" \".join(glob2.glob(genome_folder + \"*.fa*\"))\n",
    "gtf_files = \" \".join(glob2.glob(genome_folder + \"*.gtf*\"))\n",
    "fastq_files = \" \".join(glob2.glob(fastq_folder + \"*.fastq*\"))\n",
    "\n",
    "\n",
    "# +--------------------------------------------------+\n",
    "# The function for messages\n",
    "# +--------------------------------------------------+\n",
    "def msg_output(text):\n",
    "    dash = \"-\"*len(text)\n",
    "    dash = \"-\"*70 if len(dash) > 70 else \"-\"*len(dash)\n",
    "    msg_txt = \"\\n# +\" + dash + \"+\\n> {}\\n# +\" + dash + \"+\\n\"\n",
    "    print(msg_txt.format(text))\n",
    "    \n",
    "\n",
    "# +--------------------------------------------------+\n",
    "# Execute and track the shell commands\n",
    "# +--------------------------------------------------+\n",
    "def run_command(command):\n",
    "    try:\n",
    "        return subprocess.check_output(command, shell=True)\n",
    "    except (Exception, TypeError):\n",
    "        msg_output(\"! Error!: Your command was:\\n\\t\" + command)\n",
    "\n",
    "        \n",
    "# +--------------------------------------------------+\n",
    "# Execute and track the shell commands\n",
    "# +--------------------------------------------------+\n",
    "def confirmation_runCommand(command):\n",
    "    msg_output(\"Your command is:\\n\\t\" + command)\n",
    "    qa = input(\"> Are you OK with that command? Type 'YES' or 'NO': \")\n",
    "    if qa.upper() == \"YES\":\n",
    "        output = run_command(command).decode(\"utf-8\")\n",
    "        msg_output(output)\n",
    "    elif qa.upper() == \"NO\":\n",
    "        print(\"! You can change the command and then, re-run the cell\")\n",
    "    else:\n",
    "        print(\"! Just type YES or NO: Please, re-run the cell\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quality Control using FastQC\n",
    "##### Website: https://www.bioinformatics.babraham.ac.uk/projects/fastqc/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastqc_command = \"{}fastqc {} -f fastq -o {}\".format(FastQC_bin, fastq_files, results_folder + \"QC_reports\")\n",
    "confirmation_runCommand(fastqc_command)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Adapter Trimming using cutadapt\n",
    "##### Website: https://cutadapt.readthedocs.io/en/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "preprocessing_ans = input(\"> Are the adapter sequnces of your FASTQ files trimmed? 'YES' or 'NO' : \")\n",
    "\n",
    "if preprocessing_ans.upper() == \"YES\":\n",
    "    msg_output(\"! The process was terminaled because the adapter sequences already trimmed.\") \n",
    "elif preprocessing_ans.upper() == \"NO\":\n",
    "    adapter_seq = input(\"> Paste your adapter sequence: \")\n",
    "    number_of_threads = input(\"> Number Of Threads: \")\n",
    "    if True == number_of_threads.isdigit() == adapter_seq.isalpha():\n",
    "        for fastq_file in fastq_files.split(\" \"):\n",
    "            fastq_file_name = fastq_file.split(\"\\\\\")[1]\n",
    "            cutadapt_command = \"cutadapt -a {} -j {} {} -o {}trimmed_{}\"\\\n",
    "                                .format(adapter_seq, number_of_threads, fastq_file, fastq_folder, fastq_file_name)\n",
    "            confirmation_runCommand(cutadapt_command)\n",
    "    else:\n",
    "        msg_output(\"! Check the number of threads or the adapter sequence you have typed!\")\n",
    "else:\n",
    "    msg_output(\"! Type only 'YES' or 'NO'.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Curation of Genome Index using BowTie2\n",
    "#### Website: http://bowtie-bio.sourceforge.net/bowtie2/index.shtml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bowtie_base_name = input(\"Type a basename for the files (e.g.: speciesName): \")\n",
    "number_of_threads = input(\"> Number Of Threads: \")\n",
    "extra_option = input(\"> Type your extra options: \\n Check manual from http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml: \\n\")\n",
    "\n",
    "if True == number_of_threads.isdigit():\n",
    "    bowtie_build_command = \"{}bowtie2-build --threads {} {} {} {}\"\\\n",
    "                            .format(bowtie_bin, number_of_threads, fasta_files, index_folder + bowtie_base_name, extra_option)\n",
    "    confirmation_runCommand(bowtie_build_command)\n",
    "    # To check your index, use following command: bowtie2-inspect -s <base_name>\n",
    "else:\n",
    "    msg_output(\"! Check the number of threads you have typed!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [Alternative] : Curation of Genome Index using STAR\n",
    "##### Website: https://github.com/alexdobin/STAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_mode = \"genomeGenerate\"\n",
    "number_of_threads = input(\"> Number Of Threads: \")\n",
    "overhang_number = input(\"> Overhang (ideally: ReadLength - 1): \")\n",
    "extra_option = input(\"> Paste your extra options: \\nUse formal manual: https://raw.githubusercontent.com/alexdobin/STAR/921a50b1b4730a2c8b6bffc03b85081e9de3f777/doc/STARmanual.pdf \\nExample: --limitSjdbInsertNsj 4000 --limitGenomeGenerateRAM 269860224 --genomeSAindexNbases 12\\n\")\n",
    "\n",
    "if True == number_of_threads.isdigit() == overhang_number.isdigit():\n",
    "    if len(glob2.glob(genome_folder+\"*.fasta\")) > 0:\n",
    "        run_command(\"gzip {}\".format(genome_folder + \"*fasta\")).decode(\"utf-8\")\n",
    "    indexing_command = \"{}STAR --runThreadN {} --runMode {} --genomeDir {} --genomeFastaFiles {} --sjdbGTFfile {} --sjdbOverhang {} {}\" \\\n",
    "                        .format(STAR_bin, number_of_threads, run_mode, index_folder, fasta_files, gtf_files, overhang_number, extra_option)\n",
    "    confirmation_runCommand(indexing_command)\n",
    "else:\n",
    "    msg_output(\"! Check the number of threads and overhang number you have typed!\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping/Alignment using TopHat\n",
    "##### Website: http://ccb.jhu.edu/software/tophat/index.shtml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "library_type = input(\"> Library type 'fr-unstranded', ' fr-firststrand' or 'fr-secondstrand' : \")\n",
    "number_of_threads = input(\"> Number Of Threads: \")\n",
    "extra_option = input(\"> Type your extra options: \\n Check manual from http://ccb.jhu.edu/software/tophat/manual.shtml#toph: \\n\")\n",
    "msg_output(\"Please note that it is highly recommended that a FASTA file with the sequence(s) the genome being indexed be present \\n in the same directory with the Bowtie index files and having the name <genome_index_base>.fa. \\nIf not present, TopHat will automatically rebuild this FASTA file from the Bowtie index files.\")\n",
    "\n",
    "if True == number_of_threads.isdigit():\n",
    "    # TO-DO: Find an elegant way to handle that problem with regex!\n",
    "    reading_before_names = []\n",
    "    # To take only file names of FASTQ files\n",
    "    for fastq_file in fastq_files.split(\" \"):\n",
    "        reading_before_names.append(fastq_file.split(\"_\")[0].split(\"\\\\\")[1])\n",
    "    for read_file in list(set(reading_before_names)):\n",
    "        read_files_together = \",\".join(glob2.glob(fastq_folder + read_file + \"*\"))\n",
    "        tophat_command = \"{}tophat2 -p {} -o {} --library-type {} -G {} {} {} {}\"\\\n",
    "                        .format(TopHat_bin, number_of_threads, bam_sam_folder + read_file, library_type, \n",
    "                                gtf_files, index_folder + \"bowtie_base_name\", read_files_together, extra_option)\n",
    "        confirmation_runCommand(tophat_command)\n",
    "else:\n",
    "    msg_output(\"! Check the number of threads you have typed!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [Alternative]: Mapping/Alignment using STAR\n",
    "##### Website: https://github.com/alexdobin/STAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "number_of_threads = input(\"> Number Of Threads: \")\n",
    "extra_option = input(\"> Type your extra options: \\nUse formal manual: https://github.com/alexdobin/STAR\\nExample: --outSAMunmapped Within --outSAMattributes Standard\\n\")\n",
    "\n",
    "if True == number_of_threads.isdigit():\n",
    "    reading_before_names = []\n",
    "    # To take only file names of FASTQ files\n",
    "    for fastq_file in fastq_files.split(\" \"):\n",
    "        reading_before_names.append(fastq_file.split(\"_\")[0].split(\"\\\\\")[1])\n",
    "    for read_file in list(set(reading_before_names)):\n",
    "        read_files_together = \",\".join(glob2.glob(fastq_folder + read_file + \"*\"))\n",
    "        star_command = \"{}STAR --runThreadN {} --genomeDir {} --readFilesIn {} --outFileNamePrefix {} --readFilesCommand zcat --outSAMtype BAM SortedByCoordinate {}\" \\\n",
    "                .format(STAR_bin, number_of_threads, index_folder, read_files_together, bam_sam_folder + read_file, extra_option)\n",
    "        confirmation_runCommand(star_command)\n",
    "else:\n",
    "    msg_output(\"! Check the number of threads you have typed!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Index BAM files (.BAI) using samtools\n",
    "##### Website: http://www.htslib.org/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bam_files = glob2.glob(bam_sam_folder + \"*.bam\")\n",
    "bai_files = [bam_file + \".bai\" for bam_file in bam_files]\n",
    "for i in range(len(bam_files)):\n",
    "    bai_command = \"samtools index {} {}\".format(bam_files[i], bai_files[i])\n",
    "    confirmation_runCommand(bai_command)\n",
    "\n",
    "msg_output(\"! Your .BAI and .BAM files are stored in the 'bam_sam' folder.\\n You can visualize them using IGV.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Counting reads using HTSeq\n",
    "##### Website: https://htseq.readthedocs.io/en/latest/count.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mode = input(\"> Choose a mode from 'union', 'intersection-strict' or 'intersection-nonempty' :\")\n",
    "stranded = input(\"> Data is stranded? 'yes', 'reverse' or 'no' : \")\n",
    "order = input(\"> How the input data has been sorted? 'name' or 'pos' : \")\n",
    "id_attribute = input(\"> Choose an id attribute? e.g: 'gene_id'\")\n",
    "feature_type = input(\"> Feature type (3rd column in GFF file) to be used? e.g: 'exon' : \")\n",
    "extra_option = input(\"> Paste your extra options: e.g: --additional-attr=gene_name : \")\n",
    "\n",
    "for bam_file in bam_files:\n",
    "    output_fn = results_folder + \"counts/\" + bam_file.split(\"\\\\\")[-1] + \"_HTSeq.txt\"\n",
    "    htseq_command = \"htseq-count -f bam -m {} -s {} -r {} -i {} -t {} {} {} {} > {}\"\\\n",
    "                    .format(mode, stranded, order, id_attribute, feature_type, extra_option, bam_file, gtf_files, output_fn)\n",
    "    confirmation_runCommand(htseq_command)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### [Alternative]: Counting reads using featureCounts\n",
    "##### Website: http://subread.sourceforge.net/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "number_of_threads = input(\"> Number Of Threads: \")\n",
    "stranded = input(\"> Data is stranded? 'yes', 'reverse' or 'no' : \")\n",
    "id_attribute = input(\"> Choose an id attribute? e.g: 'gene_id'\")\n",
    "feature_type = input(\"> Feature type (3rd column in GFF file) to be used? e.g: 'exon' : \")\n",
    "extra_option = input(\"> Type your extra options: (Check manual from http://bioinf.wehi.edu.au/featureCounts/)\\n \")\n",
    "\n",
    "# featureCount strand information\n",
    "strand_conversion = {\"no\" : 0, \"yes\" : 1, \"reverse\" : 2}\n",
    "try:\n",
    "    stranded = strand_conversion[xz.lower()]\n",
    "except Exception:\n",
    "    msg_output(\"! Data is stranded? Select one of those: 'yes', 'reverse' or 'no' .\")\n",
    "\n",
    "if True == number_of_threads.isdigit():\n",
    "    for bam_file in bam_files:\n",
    "        output_fn = results_folder + \"counts/\" + bam_file.split(\"\\\\\")[-1] + \"_featureCounts.txt\"\n",
    "        featureCounts_command = \"featureCounts -T {} -t {} -g {} -s {} {} -a {} -o {} {}\"\\\n",
    "                        .format(number_of_threads, feature_type, id_attribute, stranded, \n",
    "                                extra_option, gtf_files, output_fn, bam_file)        \n",
    "        confirmation_runCommand(featureCounts_command)\n",
    "else:\n",
    "    msg_output(\"! Check the number of threads you have typed!\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Creating meta data file for the files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ms_presence = input(\"Do you have a meta data with a full of comma seperated values? : 'YES' or 'NO' : \")\n",
    "counts_files = [temp_file.split(\"\\\\\")[1].split(\"_\")[0] for temp_file in glob2.glob(results_folder + \"counts/*.txt\")]\n",
    "\n",
    "print(counts_files)\n",
    "if ms_presence.upper() == \"YES\":\n",
    "    msg_output(\"! Your meta data file must be in the 'results' folder as 'meta_data.csv'\\n\\twith a full of comma seperated values including 'id,condition' header.\")\n",
    "elif ms_presence.upper() == \"NO\":\n",
    "    meta_data_table = {\"id\" : \"condition\"}\n",
    "    for count_file in counts_files:\n",
    "        condition = input(\"> What is the condition for {} such as 'Control' or 'XYZ_gene mutant' ?\".format(count_file))\n",
    "        meta_data_table[count_file] = condition\n",
    "        with open(results_folder + \"meta_data.csv\", \"w\") as meta_data_file:\n",
    "            for meta_data_key in meta_data_table:\n",
    "                meta_data_file.write(meta_data_key + \",\" + meta_data_table[meta_data_key] + \"\\n\")\n",
    "    msg_output(\"! Your meta data file has been created in the 'results' folder.\")\n",
    "else:\n",
    "    msg_output(\"! Please, type either 'YES' or 'NO'.\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Differential Expression Analysis using DESeq2\n",
    "##### Website: https://bioconductor.org/packages/DESeq2"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
