{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# RNA-Seq Workflow by @furkanmtorun \n",
    "### [furkanmtorun@gmail.com](mailto:furkanmtorun@gmail.com)  | GitHub: [@furkanmtorun](https://github.com/furkanmtorun)  | [Google Scholar](https://scholar.google.com/citations?user=d5ZyOZ4AAAAJ) | [Personal Website](https://furkanmtorun.github.io/) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Libraries , packages and required functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# +--------------------------------------------------+\n",
    "# Import required librarys & packages\n",
    "# +--------------------------------------------------+\n",
    "import glob2\n",
    "import subprocess\n",
    "\n",
    "# +--------------------------------------------------+\n",
    "# Define folders and bin for tools\n",
    "# +--------------------------------------------------+\n",
    "fastq_folder, genome_folder, index_folder, bam_sam_folder, logs_folder, results_folder, \\\n",
    "FastQC_bin, STAR_bin, cufflinks_bin, bowtie_bin, TopHat_bin, R_bin = [\"./files/fastq/\", \n",
    "            \"./files/genome/\", \"./files/index/\", \"./files/bam_sam/\", \"./files/logs/\", \"./files/results/\", \"./softs/FastQC/\", \n",
    "            \"./softs/STAR-2.7.3a/bin/Linux_x86_64/\", \"./softs/cufflinks-2.2.1.Linux_x86_64/\", \n",
    "            \"./softs/bowtie2-2.3.5.1-linux-x86_64/\", \"./softs/tophat-2.1.1.Linux_x86_64/\", \n",
    "            \"C:/Program Files/R/R-3.6.1/bin/i386/\"]\n",
    "\n",
    "\n",
    "# +--------------------------------------------------+\n",
    "# Define files\n",
    "# +--------------------------------------------------+\n",
    "fasta_files = \" \".join(glob2.glob(genome_folder + \"*.fa*\"))\n",
    "gtf_files = \" \".join(glob2.glob(genome_folder + \"*.gtf*\"))\n",
    "fastq_files = \" \".join(glob2.glob(fastq_folder + \"*.fastq*\"))\n",
    "\n",
    "\n",
    "# +--------------------------------------------------+\n",
    "# The function for messages\n",
    "# +--------------------------------------------------+\n",
    "def msg_output(text):\n",
    "    msg_txt = \"\\n# +\" + \"-\"*len(text) + \"+\\n> {}\\n# +\" + \"-\"*len(text) + \"+\\n\"\n",
    "    print(msg_txt.format(text))\n",
    "    \n",
    "\n",
    "# +--------------------------------------------------+\n",
    "# Execute and track the shell commands\n",
    "# +--------------------------------------------------+\n",
    "def run_command(command):\n",
    "    try:\n",
    "        return subprocess.check_output(command, shell=True)\n",
    "    except (Exception, TypeError):\n",
    "        msg_output(\"! Error!: Your command was:\\n\\t\" + command)\n",
    "\n",
    "        \n",
    "# +--------------------------------------------------+\n",
    "# Execute and track the shell commands\n",
    "# +--------------------------------------------------+\n",
    "def confirmation_runCommand(command):\n",
    "    msg_output(\"Your command is:\\n\\t\" + command)\n",
    "    qa = input(\"> Are you OK with that command? Type 'YES' or 'NO': \")\n",
    "    if qa.upper() == \"YES\":\n",
    "        output = run_command(command).decode(\"utf-8\")\n",
    "        msg_output(output)\n",
    "    elif qa.upper() == \"NO\":\n",
    "        print(\"! You can change the command and then, re-run the cell\")\n",
    "    else:\n",
    "        print(\"! Just type YES or NO: Please, re-run the cell\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Quality Control using FastQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fastqc_command = \"{}fastqc {}*.fastq* -f fastq -o {}\".format(FastQC_bin, fastq_folder, results_folder + \"QC_reports\")\n",
    "confirmation_runCommand(fastqc_command)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Curation of Genome Index using STAR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "run_mode = \"genomeGenerate\"\n",
    "number_of_threads = input(\"> Number Of Threads: \")\n",
    "overhang_number = input(\"> Overhang (ideally: ReadLength - 1): \")\n",
    "extra_option = input(\"> Paste your extra options: \\nUse formal manual: https://raw.githubusercontent.com/alexdobin/STAR/921a50b1b4730a2c8b6bffc03b85081e9de3f777/doc/STARmanual.pdf \\nExample: --limitSjdbInsertNsj 4000 --limitGenomeGenerateRAM 269860224 --genomeSAindexNbases 12\\n\")\n",
    "if True == number_of_threads.isdigit() and True == overhang_number.isdigit():\n",
    "    if len(glob2.glob(genome_folder+\"*.fasta\")) > 0:\n",
    "        run_command(\"gzip {}\".format(genome_folder + \"*fasta\")).decode(\"utf-8\")\n",
    "    indexing_command = \"{}STAR --runThreadN {} --runMode {} --genomeDir {} --genomeFastaFiles {} --sjdbGTFfile {} --sjdbOverhang {} {}\" \\\n",
    "                        .format(STAR_bin, number_of_threads, run_mode, index_folder, fasta_files, gtf_files, overhang_number, extra_option)\n",
    "    confirmation_runCommand(indexing_command)\n",
    "else:\n",
    "    msg_output(\"Check the number of threads and overhang number you have typed!\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
